pfcp/adapter/adapter.go

```diff
@@ -257,8 +238,64 @@ func HandlePfcpSessionEstablishmentResponse(msg *udp.Message) {
 	}
 
 	// Get N3 interface UPF
+	defaultPath := smContext.Tunnel.DataPathPool.GetDefaultPath()
+	if defaultPath == nil {
+		logger.PfcpLog.Errorln("failed to get default path")
+		return
+	}
 	ANUPF := smContext.Tunnel.DataPathPool.GetDefaultPath().FirstDPNode
 
+	if rsp.CreatedPDR != nil {
+		ueIPAddress := FindUEIPAddress(rsp.CreatedPDR)
+		if ueIPAddress != nil {
+			smContext.SubPfcpLog.Infof("upf provided ue ip address [%v]", ueIPAddress)
```

Comment:
Does UPF returns SMF allocated address or it returns separate address?

Reply:
The UPF can be configured to allocate IP addresses. The UPF would specify that it is responsible for allocating IP addresses via the PFCP association setup response sent to the SMF.  When that's the case, the UPF is entirely responsible for UE IP allocation.